File: 0053_maximum_subarray.py
# https://leetcode-cn.com/problems/maximum-subarray/

# Let res[i] denote the maximum sum of subarraies that stop at nums[i].
# Then we have:
# res[0] = nums[0],
# res[1] = max(res[0]+nums[1], nums[1]),
# res[2] = max(res[1]+nums[2], nums[2])....
# Finally, we extract the maximum value in res.

class Solution:
    def maxSubArray(self, nums:List[int]) -> int:
        res = [nums[0]]
        for i in range(1,len(nums)):
            res.append(max(res[i-1]+nums[i], nums[i]))
        return max(res)

File: Cpp/0053_maximum_subarray.cpp
// https://leetcode.cn/problems/maximum-subarray/

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int sum_till_now = nums[0], sum_till_last = nums[0];
        int res = nums[0];
        for (int i = 1; i < nums.size(); ++i) {
            sum_till_now = max(nums[i], sum_till_last + nums[i]);
            res = max(res, sum_till_now);
            sum_till_last = sum_till_now;
        }
        return res;
    }
};
